knowledge-base/adr/0031-resilience-options.md

```diff
@@ -0,0 +1,310 @@
+- Target system requests:
+  - CSRF token request to the target system for non-read requests
+  - actual request to the target system
+
+### Status Quo
+
+We have two interfaces considering resilience:
+
+```ts
+export interface ResilienceOptions {
+  enableCircuitBreaker?: boolean;
+  timeout?: number;
+}
+
+export interface HttpRequestConfigBase {
+  timeout?: number | false | undefined;
+}
+```
+
+- The `ResilienceOptions` are applied to the BTP service calls. They contain a circuit breaker and timeout.
+- The timeout in the `HttpRequestConfigBase` is for all http calls to the target system and passed to axios.
+
+The settings are passed in the following way:
+
+```js
+myApi.getAll().timeout(20).execute({ enableCircuitBreaker: true, timeout: 10 });
+executeHttpRequest(
+  { enableCircuitBreaker: true, timeout: 10 },
+  { timeout: 20 }
+);
+```
+
+So we have something for circuit breaker and timeout.
+
+|     Option      | On target | On BTP | Default target | Default BTP |
+| :-------------: | :-------: | :----: | :------------: | :---------: |
+| circuit breaker |    ❌     |   ✅   |      n.a.      |   enabled   |
+|     timeout     |    ✅     |   ✅   |    enabled     |   enabled   |
+|      retry      |    ❌     |   ❌   |      n.a.      |    n.a.     |
+|   rate limit    |    ❌     |   ❌   |      n.a.      |    n.a.     |
+|   bulk limit    |    ❌     |   ❌   |      n.a.      |    n.a.     |
+
+## Decision
+
+- Rate limit and bulk limit we will not be done -> no user request.
+- We will implement an [extendable middleware approach]('#variant-c').
+- Step 1: Circuit breaker is in resilience middleware and made tenant aware
+- Step 2: Retry is in resilience middleware and excluded if circuit breaker is open.
+- Step 3 (Optional): Make resilience globally configurable for all requests.
+  Do this on demand or after customer feedback.
+- Default for circuit breaker is on.
+- Default for retry is off since this is a bigger behavior change
+
+|     Option      | On target | On BTP | Default target | Default BTP | Remarks                                    |
+| :-------------: | :-------: | :----: | :------------: | :---------: | ------------------------------------------ |
+| circuit breaker |    ✅     |   ✅   |    enabled     |   enabled   | tenant aware                               |
+|     timeout     |    ✅     |   ✅   |    enabled     |   enabled   |                                            |
+|      retry      |    ✅     |   ✅   |    disabled    |  disabled   | no retry: circuit breaker open and 401,403 |
+|   rate limit    |    ❌     |   ❌   |      n.a.      |    n.a.     |                                            |
+|   bulk limit    |    ❌     |   ❌   |      n.a.      |    n.a.     |                                            |
+
+## Consequences
+
+The user has the option to switch-on and adjust resilience.
+If options are not sufficient a custom implementation can be used.
```

Comment:
```suggestion
If options are not sufficient, custom implementations can be used.
```

Reply:
If we use the "enabled by default", we will probably make the resilience middleware a dependency and therefore always include the third party libs even if they are not needed.

Alternatively they could be a peerDependency and replaced with something else (allowing for alterntaive libs or no libs). This is probably too complex though.

Reply:
The opsuum is already in and async-retry less than 100kb install size so we thought it is ok to include it even if not used.